### DISCUSSION TITLE
Added Coral TPU to diffbot

### DISCUSSION THREAD

**Discussion by joeuser846** [fjp/2wd-robot#62] [General]
FYI, I recently added a USB [camera](https://www.amazon.com/gp/product/B0861M62KW/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) and Coral USB TPU to my diffbot. [This](https://github.com/jsk-ros-pkg/coral_usb_ros) library provides excellent Coral TPU to ROS integration, is very well documented, and the author is very helpful if problems arise. I found that my RPi 4 got overloaded if I ran the diffbot code+USB camera+IMU driver+IMU sensor fusion+TPU so I moved the TPU and related library to the laptop that runs the ROS navigation code.  

I can now drive/navigate my diffbot with live (front view) video in rviz.  I can also feed that video to the coral-usb-ros library and it will detect common household objects and box them in the rviz video window.

**Comment by fjp** [comment]
@joeuser846 sounds great, thanks for sharing your experience with the Coral TPU. 

I hope to test this as well and document the resource issues when running different things on the Pi. Now that you mention such issues, I should add a section to the docs which covers htop to monitor the load.

**Reply by joeuser846** [reply]
My RPi 4 was spiking above 90% CPU utilization when running all of the processes listed above.  The main impact was a high error rate on the RPi to Teensy interface.  That resulted in loss of reliable control of the motors.  It might help to move the IMU driver to the Teensy as you plan.  Although, that would still leave sensor fusion (robot_localization) and the Coral driver (coral_usb_ros) on the SBC so it's hard to say if that would be OK.  Given that the navigation logic resides on the laptop anyway I decided that moving object detection there too wasn't unreasonable.
